### PR TITLE
refactor: ensure that we only have a partition ID ONCE within the pipeline

### PR DESCRIPTION
**Pull request by crepererum**
This is not strictly needed at the moment but will:

- increase resilience
- help w/ #6750
- demonstrate a concept that I wanna use to actually address #6727

### CODE REVIEW

File: compactor2/src/components/combos/unique_partitions.rs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+use async_trait::async_trait;
+use data_types::PartitionId;
+
+use crate::components::{
+    partition_done_sink::PartitionDoneSink, partitions_source::PartitionsSource,
+};
+
+pub fn unique_partitions<T1, T2>(
+    inner_source: T1,
+    inner_sink: T2,
```
Comment (NGA-TRAN) [inline compactor2/src/components/combos/unique_partitions.rs:18]: It is so hard to follow. I wonder if you add comments `briefly` explain the purpose of these 2 and how it work will help. Without this if we hit incident, it will be hard to read to code and figure out whether it introduce bug or not

**Review by NGA-TRAN — COMMENTED** [review]
@crepererum 
I will continue reviewing this after you provide some comments

File: compactor2/src/components/combos/unique_partitions.rs (hunk 1 above)
Comment (crepererum) [reply compactor2/src/components/combos/unique_partitions.rs:38]: done

Comment (crepererum) [reply compactor2/src/components/combos/unique_partitions.rs:38]: image

File: compactor2/src/components/combos/unique_partitions.rs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+        inner_source.set(vec![
+            PartitionId::new(1),
+            PartitionId::new(3),
+            PartitionId::new(5),
+        ]);
+
+        assert_eq!(
+            source.fetch().await,
+            vec![PartitionId::new(1), PartitionId::new(5)],
+        );
```
Comment (NGA-TRAN) [inline compactor2/src/components/combos/unique_partitions.rs:201]: OK, I go tit now. You start with a list of duplicates `1, 1, 2, 3, 4`. Then you only fetch the unique `1, 2, 3, 4`. Then you remove `1, 2` which means we did fetch only `3, 4`. So when you set your source to `1, 3, 5`, you should only need to fetch `1, 5` because `3` was fetched and stay 👍

**Review by NGA-TRAN — APPROVED** [review]
Got the idea and verified the tests
